internal/api/creator.go

```diff
@@ -0,0 +1,200 @@
+	replaceTemplates(mld *ModuleLoaderData) error
+}
+
+type moduleLoaderDataCreatorHelperImpl struct {
+	sectionGetter sectionGetter
+}
+
+func (kh *moduleLoaderDataCreatorHelperImpl) findKernelMapping(mappings []kmmv1beta1.KernelMapping, kernelVersion string) (*kmmv1beta1.KernelMapping, error) {
+	for _, m := range mappings {
+		if m.Literal != "" && m.Literal == kernelVersion {
+			return &m, nil
+		}
+
+		if m.Regexp == "" {
+			continue
+		}
+
+		if matches, err := regexp.MatchString(m.Regexp, kernelVersion); err != nil {
+			return nil, fmt.Errorf("could not match regexp %q against kernel %q: %v", m.Regexp, kernelVersion, err)
+		} else if matches {
+			return &m, nil
+		}
+	}
+
+	return nil, errors.New("no suitable mapping found")
+}
+
+func (kh *moduleLoaderDataCreatorHelperImpl) prepareModuleLoaderData(mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (*ModuleLoaderData, error) {
+	var err error
+
+	mld := &ModuleLoaderData{}
+	// prepare the build
+	if mapping.Build != nil || mod.Spec.ModuleLoader.Container.Build != nil {
+		mld.Build = kh.sectionGetter.GetRelevantBuild(mod.Spec.ModuleLoader.Container.Build, mapping.Build)
+	}
+
+	// prepare the sign
+	if mapping.Sign != nil || mod.Spec.ModuleLoader.Container.Sign != nil {
+		mld.Sign, err = kh.sectionGetter.GetRelevantSign(mod.Spec.ModuleLoader.Container.Sign, mapping.Sign, kernelVersion)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get the relevant Sign configuration for kernel %s: %v", kernelVersion, err)
+		}
+	}
+
+	// prepare TLS options
+	mld.RegistryTLS = mapping.RegistryTLS
+	if mapping.RegistryTLS == nil {
+		mld.RegistryTLS = &mod.Spec.ModuleLoader.Container.RegistryTLS
+	}
+
+	//prepare container image
+	mld.ContainerImage = mapping.ContainerImage
+	if mapping.ContainerImage == "" {
+		mld.ContainerImage = mod.Spec.ModuleLoader.Container.ContainerImage
+	}
+
+	mld.KernelVersion = kernelVersion
+	mld.Name = mod.Name
+	mld.Namespace = mod.Namespace
+	mld.ImagePullPolicy = mod.Spec.ModuleLoader.Container.ImagePullPolicy
+	mld.ImageRepoSecret = mod.Spec.ImageRepoSecret
+	mld.Selector = mod.Spec.Selector
+	mld.ServiceAccountName = mod.Spec.ModuleLoader.ServiceAccountName
+	mld.Modprobe = mod.Spec.ModuleLoader.Container.Modprobe
+	mld.Owner = mod
+
+	return mld, nil
+}
+
+func (kh *moduleLoaderDataCreatorHelperImpl) replaceTemplates(mld *ModuleLoaderData) error {
+	osConfigEnvVars := utils.KernelComponentsAsEnvVars(mld.KernelVersion)
+	osConfigEnvVars = append(osConfigEnvVars, "MOD_NAME="+mld.Name, "MOD_NAMESPACE="+mld.Namespace)
+
+	replacedContainerImage, err := utils.ReplaceInTemplates(osConfigEnvVars, mld.ContainerImage)
+	if err != nil {
+		return fmt.Errorf("failed to substitute templates in the ContainerImage field: %v", err)
+	}
+	mld.ContainerImage = replacedContainerImage[0]
+
+	return nil
+}
+
+type sectionGetter interface {
+	GetRelevantBuild(moduleBuild *kmmv1beta1.Build, mappingBuild *kmmv1beta1.Build) *kmmv1beta1.Build
```

Comment:
those interface's methods can be private, they are not used anywhere outside of the package